gateway/ln-gateway/src/lib.rs

```diff
@@ -551,8 +551,17 @@ impl Gateway {
                         break;
                     };
 
-                    let htlc_request = match stream.next().await {
-                        Some(Ok(htlc_request)) => htlc_request,
+                    let payment_request = tokio::select! {
+                        payment_request = stream.next() => {
+                            payment_request
+                        }
+                        () = self.is_shutting_down_safely() => {
+                            break;
+                        }
+                    };
```

Comment:
Not sure if we fully fix raciness, but at least the time where it might happen becomes much shorter. If we start shutdown after the select, but before calling `handle_htlc`, could that still be a problem?

Reply:
Currently according to `ILnRpcClient` documentation, not calling `ILnRpcClient::complete_htlc()` for an item sent to the `ILnRpcClient::route_htlcs()` stream is undefined behavior. The underlying behavior for ldk, lnd, and cln is to fail the HTLC after the CLTV expiry time has passed. Definitely not ideal.

@m1sterc001guy can we add a TODO comment in here stating the issue above? This should definitely be addressed down the road.

Also, would it be worth defining the above behavior in `ILnRpcClient`? If so, I can make a PR for that.

Reply:
> If we start shutdown after the select, but before calling handle_htlc, could that still be a problem?

Ah yes, this is still an issue. I think this is fixable by having the section of code after the `select!` protected by a lock, and also the `handle_shutdown_msg` function protected by the same lock. That way, the gateway's state cannot change after we've checked it and call `handle_htlc`

Reply:
@elsirion I believe I fixed the remaining issue by holding the `GatewayState` RW Lock. The `handle_shutdown_msg` thread will hold the write lock and this thread will hold the read lock. Here, we now check the state after taking the lock, meaning if we are still in the `Running` state that the state should not change until `handle_htlc` is called (should be atomic now).

Thanks for double checking this!

Reply:
Looks good to me now!